x/oracle/keeper/native_token.go

```diff
@@ -0,0 +1,291 @@
+			if vPubkey == validatorPubkey {
+				stakerInfo.ValidatorPubkeyList = append(stakerInfo.ValidatorPubkeyList[:i], stakerInfo.ValidatorPubkeyList[i+1:]...)
+				break
+			}
+		}
+	}
+
+	newBalance.Balance += amount.Int64()
+
+	keyStakerList := types.NativeTokenStakerListKey(assetID)
+	valueStakerList := store.Get(keyStakerList)
+	stakerList.StakerAddrs = make([]string, 0, 1)
+	if valueStakerList != nil {
+		k.cdc.MustUnmarshal(valueStakerList, &stakerList)
+	}
+	exists := false
+	for idx, stakerExists := range stakerList.StakerAddrs {
+		// this should noly happen when do withdraw
+		if stakerExists == stakerAddr {
+			if newBalance.Balance <= 0 {
+				stakerList.StakerAddrs = append(stakerList.StakerAddrs[:idx], stakerList.StakerAddrs[idx+1:]...)
+				valueStakerList = k.cdc.MustMarshal(&stakerList)
+				store.Set(keyStakerList, valueStakerList)
+			}
+			exists = true
+			stakerInfo.StakerIndex = int64(idx)
+			break
+		}
+	}
+	if !exists {
+		if !amount.IsPositive() {
+			return errors.New("remove unexist validator")
+		}
+		stakerList.StakerAddrs = append(stakerList.StakerAddrs, stakerAddr)
+		valueStakerList = k.cdc.MustMarshal(&stakerList)
+		store.Set(keyStakerList, valueStakerList)
+		stakerInfo.StakerIndex = int64(len(stakerList.StakerAddrs) - 1)
+	}
+
+	if newBalance.Balance <= 0 {
+		store.Delete(key)
+	} else {
+		stakerInfo.BalanceList = append(stakerInfo.BalanceList, &newBalance)
+		bz := k.cdc.MustMarshal(stakerInfo)
+		store.Set(key, bz)
+	}
+	eventValue := fmt.Sprintf("%d_%s", stakerInfo.StakerIndex, validatorPubkey)
+	if newBalance.Change == types.BalanceInfo_ACTION_DEPOSIT {
+		eventValue = fmt.Sprintf("%s_%s", types.AttributeValueNativeTokenDeposit, eventValue)
+	} else {
+		eventValue = fmt.Sprintf("%s_%s", types.AttributeValueNativeTokenWithdraw, eventValue)
+	}
+	// emit an event to tell a new valdiator added/or a validator is removed for the staker
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCreatePrice,
+		sdk.NewAttribute(types.AttributeKeyNativeTokenChange, eventValue),
+	))
+	return nil
+}
+
+// GetStakerInfos returns all stakers information
+func (k Keeper) GetStakerInfos(ctx sdk.Context, assetID string) (ret []*types.StakerInfo) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.NativeTokenStakerKeyPrefix(assetID))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		sInfo := types.StakerInfo{}
+		k.cdc.MustUnmarshal(iterator.Value(), &sInfo)
+		// keep only the latest effective-balance
+		sInfo.BalanceList = sInfo.BalanceList[:len(sInfo.BalanceList)-1]
+		// this is mainly used by price feeder, so we remove the stakerAddr to reduce the size of return value
+		sInfo.StakerAddr = ""
+		ret = append(ret, &sInfo)
+	}
+	return ret
+}
+
+// GetstakerList return stakerList for native-restaking asset of assetID
+func (k Keeper) GetStakerList(ctx sdk.Context, assetID string) types.StakerList {
```

Comment:
Consider returning an additional boolean value in the **GetStakerInfo** and **GetStakerList** functions to indicate whether the data has been found

> func (k Keeper) GetStakerInfo(ctx sdk.Context, assetID, stakerAddr string) (types.StakerInfo, bool) {
    // ...
    if value == nil {
        return stakerInfo, false
    }
    k.cdc.MustUnmarshal(value, &stakerInfo)
    return stakerInfo, true
}

Reply:
when not found, it will return empty struct

Reply:
stakerList is a wrapper of slice, so could be check with length